util/mathutil/exponential_average.go

```diff
@@ -0,0 +1,67 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mathutil
+
+import "sync"
+
+func factor(n int) float64 {
+	return 2.0 / float64(n+1)
+}
+
+// ExponentialMovingAverage is an exponential moving average measurement implementation.
+type ExponentialMovingAverage struct {
+	value        float64
+	sum          float64
+	factor       int
+	warmupWindow int
+	count        int
+
+	mu sync.RWMutex
+}
+
+// NewExponentialAverage will create a new ExponentialMovingAverage
+func NewExponentialAverage(
+	factor int,
+	warmupWindow int,
+) *ExponentialMovingAverage {
+	if factor <= 1 {
+		panic("factor must be greater than 1")
+	}
+	return &ExponentialMovingAverage{
+		factor:       factor,
+		warmupWindow: warmupWindow,
+	}
+}
+
+// Add a single sample and update the internal state.
+func (m *ExponentialMovingAverage) Add(value float64) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if m.count < m.warmupWindow {
+		m.count++
+		m.sum += value
+		m.value = m.sum / float64(m.count)
+	} else {
+		f := factor(m.factor)
```

Comment:
sorry I mean we can expose the output of `factor` as a parameter to user rather than use `2.0 / float64(n+1)`. The parameter means "When a new value comes, it should affect the result of average in what percentage?" 1 means the result of average is 100% affected by the new value, 0.5 means 50% new value + 50% old average

Reply:
I have updated.